File: CPlusPlus/DIP/After.cpp
// MyUserService.cpp
#include <iostream>
#include <string>
#include <memory>

// Abstraction
class IDatabase {
public:
    virtual void Save(const std::string& userData) = 0;
    virtual ~IDatabase() = default;
};

// Concrete implementation
class MyDatabase : public IDatabase {
public:
    void Save(const std::string& userData) override {
        std::cout << "Saved to database: " << userData << std::endl;
    }
};

// High-level module depends on abstraction
class MyUsersService {
private:
    std::shared_ptr<IDatabase> database;

public:
    // Constructor injection
    MyUsersService(std::shared_ptr<IDatabase> db) : database(db) {}

    void RegisterUser(const std::string& userData) {
        database->Save(userData);
    }
};

File: CSharp/DIP/After.cs
// Abstraction
public interface IDatabase
{
    void Save(string userData);
}

// Low-level module implementing the abstraction
public class MyDatabase : IDatabase
{
    public void Save(string userData)
    {
        Console.WriteLine($"Saved to database: {userData}");
    }
}

// High-level module now depends on the abstraction
public class MyUsersService
{
    private IDatabase _database;

    // Dependency Injection via constructor
    public MyUsersService(IDatabase database)
    {
        _database = database;
    }

    public void RegisterUser(string userData)
    {
        _database.Save(userData);
    }
}


// IDatabase database = new Database(); // Concrete implementation
// var userService = new UsersService(database); // Injecting the dependency
// userService.RegisterUser("John Doe");
